.github/workflows/build.yml

```diff
@@ -43,54 +43,112 @@ jobs:
         arch: [amd64, arm64]
     timeout-minutes: ${{ (github.event_name == 'schedule' && 90) || ((matrix.arch == 'arm64' && 45) || 30) }}
     needs: timestamp
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 1
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Build slim image `--target discourse_dependencies`
+        run: |
+          cd image && ruby auto_build.rb base_slim_${{ matrix.arch }} --cache-to=type=local,dest=/tmp/build_deps_${{ matrix.arch }}
+
+      - name: Print cache
+        run: |
+          ls -lah /tmp/build_deps_${{ matrix.arch }}
+
+      - name: Upload artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: build_deps_${{ matrix.arch }}
+          path: /tmp/build_deps_${{ matrix.arch }}
+
+  base:
+    # `unbuntu-22.04-8core` for arch amd64 non-scheduled builds
+    # `unbuntu-22.04` for arch amd64 scheduled builds
+    # `unbuntu-22.04-8core-arm` for arch arm64 non-scheduled builds
+    # `unbuntu-22.04-2core-arm` for arch arm64 scheduled builds
+    runs-on: ubuntu-22.04${{ ((github.event_name != 'schedule') && '-8core') || (( matrix.arch == 'arm64' && '-2core' ) || '') }}${{ (matrix.arch == 'arm64') && '-arm' || '' }}
+    strategy:
+      matrix:
+        arch: [amd64, arm64]
+        discourse_branch: [main, stable]
+    timeout-minutes: ${{ (github.event_name == 'schedule' && 90) || ((matrix.arch == 'arm64' && 45) || 30) }}
+    needs: [deps, timestamp]
     env:
-      ARCH: ${{matrix.arch}}
       TIMESTAMP: ${{ needs.timestamp.outputs.timestamp }}
     steps:
       - uses: actions/checkout@v3
         with:
           fetch-depth: 1
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Download artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: build_deps_${{ matrix.arch }}
+          path: /tmp
+
+      - name: Print cache
+        run: |
+          ls -lah /tmp/build_deps_${{ matrix.arch }}
+
       - name: build slim image
         run: |
-          cd image && ruby auto_build.rb base_slim_$ARCH
+          cd image && ruby auto_build.rb base_slim_${{ matrix.discourse_branch }}_${{ matrix.arch }} --cache-from=type=local,src=/tmp/build_deps_${{ matrix.arch }}
+
       - name: tag slim images
         run: |
-          docker tag discourse/base:build_slim_$ARCH discourse/base:2.0.$TIMESTAMP-slim-$ARCH
-          docker tag discourse/base:build_slim_$ARCH discourse/base:slim-$ARCH
+          docker tag discourse/base:build_slim_${{ matrix.discourse_branch }}_${{ matrix.arch }} discourse/base:2.0.${{ env.TIMESTAMP }}-slim-${{ matrix.discourse_branch }}-${{ matrix.arch }}
```

Comment:
Why all the changes from `$ARCH` to `${{ matrix.arch }}`? IMO the more-concise environment variable syntax is much easier to read/write?

Reply:
O $ARCH is easier to read/write, it makes debugging stuff in GitHub’s output harder because the command printed is without the ENV substitution.